### PR TITLE
DependencyScanner: honor additional compiler flags in interfaces files when collecting imports

### PR DESCRIPTION
**Pull request by nkcsgexi**
Additional flags in interface files may change parsing behavior like #if
statements. We should use a fresh ASTContext with these additional
flags when parsing interface files to collect imports.

rdar://62612027

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please test asan

**Comment by nkcsgexi** [comment]
@swift-ci Please ASAN test

File: lib/FrontendTool/ScanDependencies.cpp [hunk 1]
```diff
@@ -58,13 +59,51 @@ static void findAllImportedClangModules(ASTContext &ctx, StringRef moduleName,
   }
 }
 
+struct InterfaceSubASTContextDelegate: SubASTContextDelegate {
+  bool runInSubContext(ASTContext &ctx, StringRef interfacePath,
```
Comment (DougGregor) [inline lib/FrontendTool/ScanDependencies.cpp:63]: Much of this logic is also in `ModuleInterfaceBuilder::buildSwiftModuleInternal`, although some bits of validation there are missing here. Is there a way to unify this with `buildSwiftModuleInternal`? For example, by parameterizing it with an `action` like this function has, and having the dependency scanner's action be "walk the imports" and the module builder's action be "type check and serialialize"? More sharing there would help us avoid any discrepancies with the main build.

Comment (nkcsgexi) [reply lib/FrontendTool/ScanDependencies.cpp:63]: Yeah, it seems to be feasible. The configuration setup here is mostly for parsing to be performed correctly, which doesn't require options like `SerializationOptions` to be configured. We could potentially move these additional setups to the action body of "type check and serialize".

Comment (DougGregor) [reply lib/FrontendTool/ScanDependencies.cpp:63]: I'd certainly prefer that. Anything that differs between the fast dependency scanner and the normal build process is going to be a hard-to-diagnose bug later on down the line.

**Pull request merged**